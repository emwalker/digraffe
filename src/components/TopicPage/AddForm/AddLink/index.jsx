// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import type { Relay, RepositoryType, TopicType, UserType } from 'components/types'
import upsertLinkMutation from 'mutations/upsertLinkMutation'

/* eslint jsx-a11y/label-has-for: 0 */

type Props = {
  disabled?: boolean,
  relay: Relay,
  topic: TopicType,
  viewer: UserType,
}

type State = {
  url: string,
}

class AddLink extends Component<Props, State> {
  static defaultProps = {
    disabled: true,
  }

  state = {
    url: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createLink()
  }

  get selectedRepo(): ?RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin(): ?string {
    const repo = this.selectedRepo
    if (!repo)
      return null
    return repo.organization.login
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_links',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'linkEdge',
    }]
  }

  updateUrl = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  createLink() {
    const repo = this.selectedRepo
    const repoName = repo ? repo.name : null

    upsertLinkMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        addParentTopicIds: [this.props.topic.id],
        organizationLogin: this.orgLogin,
        repositoryName: repoName,
        url: this.state.url,
      },
    )
    this.setState({ url: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-w"
          aria-label="Add a link to this topic."
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-link-url input-sm"
          disabled={this.props.disabled}
          id="create-link-url"
          onChange={this.updateUrl}
          onKeyPress={this.onKeyPress}
          placeholder="Url"
          type="url"
          value={this.state.url}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_viewer on User {
    selectedRepository {
      id
      name

      organization {
        login
      }
    }
  }

  fragment AddLink_topic on Topic {
    id
  }
`)
