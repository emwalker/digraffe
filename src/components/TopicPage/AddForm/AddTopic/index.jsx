// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import type { Relay, RepositoryType, TopicType, UserType } from 'components/types'
import upsertTopicMutation from 'mutations/upsertTopicMutation'

/* eslint jsx-a11y/label-has-for: 0 */

type Props = {
  disabled?: boolean,
  relay: Relay,
  topic: TopicType,
  viewer: UserType,
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  static defaultProps = {
    disabled: true,
  }

  state = {
    name: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createTopic()
  }

  get selectedRepo(): ?RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin(): ?string {
    const repo = this.selectedRepo
    if (!repo)
      return null
    return repo.organization.login
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_childTopics',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'topicEdge',
    }]
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  createTopic() {
    const repo = this.selectedRepo
    const repoName = repo ? repo.name : null

    upsertTopicMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        name: this.state.name,
        repositoryName: repoName,
        organizationLogin: this.orgLogin,
        topicIds: [this.props.topic.id],
      },
    )
    this.setState({ name: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <label htmlFor="create-topic-name">Add subtopic</label>
      </dt>
      <dd>
        <input
          className="form-control test-topic-name input-sm"
          disabled={this.props.disabled}
          id="create-topic-name"
          onChange={this.updateName}
          onKeyPress={this.onKeyPress}
          placeholder="Name or description"
          value={this.state.name}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddTopic, graphql`
  fragment AddTopic_viewer on User {
    selectedRepository {
      name
      organization {
        login
      }
    }
  }

  fragment AddTopic_topic on Topic {
    id
  }
`)
