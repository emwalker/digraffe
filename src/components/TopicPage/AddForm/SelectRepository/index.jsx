// @flow
import React, { Component } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'

import selectRepositoryMutation from 'mutations/selectRepositoryMutation'

/* eslint jsx-a11y/label-has-for: 0 */

type Props = {
  relay: {
    environment: Object,
  },
  viewer: {
    selectedRepository: {
      private: boolean,
    },
    repositories: {
      edges: Array<{
        selected: boolean,
        node: Object,
      }>,
    },
  },
}

class SelectRepository extends Component<Props> {
  onChange = (event) => {
    const repositoryId = event ? event.target.value : null
    selectRepositoryMutation(
      this.props.relay.environment,
      [],
      { repositoryId: repositoryId === 'placeholder' ? null : repositoryId },
    )
  }

  get repositoryEdges(): Object[] {
    return this.props.viewer.repositories.edges
  }

  get selectedId(): ?string {
    const repo = this.props.viewer.selectedRepository
    return repo ? repo.id : null
  }

  renderOption = edge => (
    <option
      key={edge.node.fullName}
      value={edge.node.id}
    >
      {edge.node.fullName}
    </option>
  )

  render = () => (
    <dl className="form-group" style={this.style}>
      <dt>
        <label htmlFor="select-repo">New links and topics added to</label>
      </dt>
      <dd>
        <select
          id="select-repo"
          className="form-select"
          aria-label="Repository"
          style={{ width: '100%' }}
          defaultValue={this.selectedId}
          onChange={this.onChange}
        >
          <option key="0" value="placeholder">Select a repository</option>
          {this.repositoryEdges.map(this.renderOption)}
        </select>
      </dd>
    </dl>
  )
}

export default createFragmentContainer(SelectRepository, graphql`
  fragment SelectRepository_viewer on User {
    selectedRepository {
      id
      isPrivate
    }

    repositories(first: 100) {
      edges {
        isSelected

        node {
          fullName
          id
        }
      }
    }
  }
`)
