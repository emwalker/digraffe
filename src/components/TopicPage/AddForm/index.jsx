// @flow
import React, { Component } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import classNames from 'classnames'

import { TopicType } from 'components/types'
import AddTopic from './AddTopic'
import AddLink from './AddLink'
import SelectRepository from './SelectRepository'
import './index.css'

type Props = {
  topic: TopicType,
  viewer: Object,
}

class AddForm extends Component<Props> {
  get className(): string {
    return classNames(
      'border',
      'rounded-1',
      'p-2',
      'mt-3',
      { 'private-repo': this.isPrivateRepo },
    )
  }

  get isPrivateRepo(): boolean {
    const repo = this.props.viewer.selectedRepository
    return repo && repo.isPrivate
  }

  render = () => (
    <form className={this.className}>
      <SelectRepository
        viewer={this.props.viewer}
      />
      <AddTopic
        topic={this.props.topic}
        viewer={this.props.viewer}
      />
      <AddLink
        topic={this.props.topic}
        viewer={this.props.viewer}
      />
    </form>
  )
}

export default createFragmentContainer(AddForm, graphql`
  fragment AddForm_viewer on User {
    selectedRepository {
      isPrivate
    }

    ...AddLink_viewer
    ...AddTopic_viewer
    ...SelectRepository_viewer
  }

  fragment AddForm_topic on Topic {
    ...AddLink_topic
    ...AddTopic_topic
  }
`)
