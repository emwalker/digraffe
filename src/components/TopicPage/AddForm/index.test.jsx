import React from 'react'
import { shallow } from 'enzyme'

import AddForm from './index'

const props = {
  topic: {
    id: '1234',
  },
  viewer: {
    selectedRepository: {
      name: null,
      organization: {
        login: 'gnusto',
      },
    },
  },
}

describe('<AddForm />', () => {
  const wrapper = shallow(<AddForm {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
