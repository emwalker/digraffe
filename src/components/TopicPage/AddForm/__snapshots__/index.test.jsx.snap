// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddForm /> renders 1`] = `
<ContainerConstructor
  componentRef={null}
  topic={
    Object {
      "id": "1234",
    }
  }
  viewer={
    Object {
      "selectedRepository": Object {
        "name": null,
        "organization": Object {
          "login": "gnusto",
        },
      },
    }
  }
/>
`;
